include/bm/bm_sim/field_lists.h

```diff
@@ -46,23 +47,42 @@ class FieldList {
     bool operator==(const field_t& other) const {
       return header == other.header && offset == other.offset;
     }
-
     bool operator!=(const field_t& other) const {
       return !(*this == other);
     }
   };
 
+  struct constant_t {
+    int value;
+    size_t nbits;
+
+    bool operator==(const constant_t& other) const {
+      return value == other.value && nbits == other.nbits;
+    }
+    bool operator!=(const constant_t& other) const {
+      return !(*this == other);
+    }
+  };
+
  public:
-  using iterator = std::vector<field_t>::iterator;
-  using const_iterator = std::vector<field_t>::const_iterator;
-  using reference = std::vector<field_t>::reference;
-  using const_reference = std::vector<field_t>::const_reference;
+  using field_list_member_t = boost::variant<field_t, constant_t>;
```

Comment:
I think the implementation choice (`boost::variant`) should be hidden from the `FieldList` public "interface" here. Instead you could provide a method which lets the target programmer apply a functor of his choice to members of the list. Something like this:
```
struct FieldListVisitor : boost::static_visitor<> {
  void operator ()(const field_t &field) { } // do nothing on purpose
  void operator ()(const constant_t &constant) { } // do nothing on purpose
};
class FieldList {
  // ...
  // public method that can be used by targets
  template <typename T>
  visit(const T &visitor) {
      static_assert(std::is_base_of(FieldListVisitor, T), "invalid visitor, must inherit from FieldListVisitor");
      std::for_each(fields.begin(), fields.end(), boost::apply_visitor(visitor));
  }
};
```

For convenience I think we can also provide a `FieldList::copy_fields_between_phvs` method (implemented with `FieldList::visit`) that could be used by `simple_switch`.
This will hide `boost::variant` to the target and remove the need for ugly `typeid` calls.